File: app/src/main/java/com/evangelidis/t_tmoviesseries/login/LoginActivity.java
package com.evangelidis.t_tmoviesseries.login;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;
import androidx.databinding.DataBindingUtil;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;

import com.evangelidis.t_tmoviesseries.MainActivity;
import com.evangelidis.t_tmoviesseries.R;
import com.evangelidis.t_tmoviesseries.databinding.ActivityLoginBinding;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;

import es.dmoral.prefs.Prefs;

import static android.view.View.INVISIBLE;
import static android.view.View.VISIBLE;
import static com.evangelidis.t_tmoviesseries.Constants.IS_LOGIN_SKIPPED;
import static com.evangelidis.t_tmoviesseries.login.FlexibleFrameLayout.ORDER_LOGIN_STATE;
import static com.evangelidis.t_tmoviesseries.login.FlexibleFrameLayout.ORDER_SIGN_UP_STATE;

public class LoginActivity extends AppCompatActivity {

    private ActivityLoginBinding binding;
    private boolean isLogin = true;

    private FirebaseAuth mAuth;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_login);

        mAuth = FirebaseAuth.getInstance();

        LoginFragment topLoginFragment = new LoginFragment();
        SignUpFragment topSignUpFragment = new SignUpFragment();

        getSupportFragmentManager().beginTransaction()
                .replace(R.id.login_fragment, topLoginFragment)
                .replace(R.id.sign_up_fragment, topSignUpFragment)
                .commit();

        binding.loginFragment.setRotation(-90);

        binding.button.setOnButtonSwitched(isLogin -> {
            binding.getRoot()
                    .setBackgroundColor(ContextCompat.getColor(
                            this,
                            isLogin ? R.color.colorPrimary : R.color.secondPage));
        });

        binding.loginFragment.setVisibility(INVISIBLE);
    }

    @Override
    public void onStart() {
        super.onStart();
        // Check if user is signed in (non-null) and update UI accordingly.
        FirebaseUser currentUser = mAuth.getCurrentUser();
        if (currentUser != null){
            startActivity(new Intent(LoginActivity.this, MainActivity.class));
            finish();
        } else if (Prefs.with(this).readBoolean(IS_LOGIN_SKIPPED, false)){
            startActivity(new Intent(LoginActivity.this, MainActivity.class));
            finish();
        }
    }

    @Override
    public void onWindowFocusChanged(boolean hasFocus) {
        super.onWindowFocusChanged(hasFocus);
        binding.loginFragment.setPivotX(binding.loginFragment.getWidth() / 2);
        binding.loginFragment.setPivotY(binding.loginFragment.getHeight());
        binding.signUpFragment.setPivotX(binding.signUpFragment.getWidth() / 2);
        binding.signUpFragment.setPivotY(binding.signUpFragment.getHeight());
    }

    public void switchFragment(View v) {
        if (isLogin) {
            binding.loginFragment.setVisibility(VISIBLE);
            binding.loginFragment.animate().rotation(0).setListener(new AnimatorListenerAdapter() {
                @Override
                public void onAnimationEnd(Animator animation) {
                    super.onAnimationEnd(animation);
                    binding.signUpFragment.setVisibility(INVISIBLE);
                    binding.signUpFragment.setRotation(90);
                    binding.wrapper.setDrawOrder(ORDER_LOGIN_STATE);
                }
            });
        } else {
            binding.signUpFragment.setVisibility(VISIBLE);
            binding.signUpFragment.animate().rotation(0).setListener(new AnimatorListenerAdapter() {
                @Override
                public void onAnimationEnd(Animator animation) {
                    super.onAnimationEnd(animation);
                    binding.loginFragment.setVisibility(INVISIBLE);
                    binding.loginFragment.setRotation(-90);
                    binding.wrapper.setDrawOrder(ORDER_SIGN_UP_STATE);
                }
            });
        }

        isLogin = !isLogin;
        binding.button.startAnimation();
    }
}

File: app/src/main/java/com/evangelidis/t_tmoviesseries/view/login/LoginActivity.kt
package com.evangelidis.t_tmoviesseries.view.login

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.content.Intent
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import com.evangelidis.t_tmoviesseries.R
import com.evangelidis.t_tmoviesseries.databinding.ActivityLoginBinding
import com.evangelidis.t_tmoviesseries.extensions.hide
import com.evangelidis.t_tmoviesseries.extensions.show
import com.evangelidis.t_tmoviesseries.utils.Constants.IS_LOGIN_SKIPPED
import com.evangelidis.t_tmoviesseries.view.main.MainActivity
import com.google.firebase.auth.FirebaseAuth
import es.dmoral.prefs.Prefs

class LoginActivity : AppCompatActivity() {

    private var isLogin = true
    private var mAuth: FirebaseAuth? = null

    private val binding: ActivityLoginBinding by lazy { ActivityLoginBinding.inflate(layoutInflater) }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(binding.root)

        mAuth = FirebaseAuth.getInstance()
        val topLoginFragment = LoginFragment()
        val topSignUpFragment = SignUpFragment()

        supportFragmentManager.beginTransaction()
            .replace(R.id.login_fragment, topLoginFragment)
            .replace(R.id.sign_up_fragment, topSignUpFragment)
            .commit()

        binding.loginFragment.rotation = -90f

        binding.button.apply {
            setOnButtonSwitched(object : OnButtonSwitchedListener {
                override fun onButtonSwitched(isLogin: Boolean) {
                    binding.loginLayout.setBackgroundColor(ContextCompat.getColor(this@LoginActivity, if (isLogin) R.color.colorPrimary else R.color.secondPage))
                }
            })
            setOnClickListener {
                switchFragment()
            }
        }

        binding.loginFragment.hide()
    }

    public override fun onStart() {
        super.onStart()
        // Check if user is signed in (non-null) and update UI accordingly.
        if ((mAuth?.currentUser != null) || (Prefs.with(this).readBoolean(IS_LOGIN_SKIPPED, false))) {
            startActivity(Intent(this@LoginActivity, MainActivity::class.java))
            finish()
        }
    }

    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)

        binding.loginFragment.apply {
            pivotX = this.width / 2.toFloat()
            pivotY = this.height.toFloat()
        }
        binding.signUpFragment.apply {
            pivotX = this.width / 2.toFloat()
            pivotY = this.height.toFloat()
        }
    }

    private fun switchFragment() {
        if (isLogin) {
            binding.loginFragment.apply {
                show()
                animate().rotation(0f).setListener(object : AnimatorListenerAdapter() {
                    override fun onAnimationEnd(animation: Animator) {
                        super.onAnimationEnd(animation)
                        binding.signUpFragment.apply {
                            hide()
                            rotation = 90f
                        }
                        binding.wrapper.setDrawOrder(FlexibleFrameLayout.ORDER_LOGIN_STATE)
                    }
                })
            }
        } else {
            binding.signUpFragment.apply {
                show()
                animate().rotation(0f).setListener(object : AnimatorListenerAdapter() {
                    override fun onAnimationEnd(animation: Animator) {
                        super.onAnimationEnd(animation)
                        binding.loginFragment.apply {
                            hide()
                            rotation = -90f
                        }
                        binding.wrapper.setDrawOrder(FlexibleFrameLayout.ORDER_SIGN_UP_STATE)
                    }
                })
            }
        }
        isLogin = !isLogin
        binding.button.startAnimation()
    }
}
